# Exlporing eBay Car Sales Data

In this project, we will explore the dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website

The goal of our project is to clean the data and analyze the used car listings.





## Data Source

Collecting data takes significant amount of time and energy. So we must always try to find relevant existing dataset.Luckily, there is a datset that is suitable for our purpose.This dataset is available on Kaggle.

[A data set](https://www.kaggle.com/orgesleka/used-cars-database/data) containing data about 370,000 used cars.

NOTE: This dataset has been reduced from 370,000 rows to 50,000 rows.



## Reading the Dataset


In [1]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding = "Latin-1") 
autos.info()
autos.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Cleaning Column Names

We can see that the column names use camelcase instead of Python's preferred snakecase. Lets reword some of the column names based on the data dictionary to be more descriptive.


In [2]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [3]:
autos.columns = (['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'no_pictures', 'postal_code',
       'last_seen'])
autos.info()
autos.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
date_crawled          50000 non-null object
name                  50000 non-null object
seller                50000 non-null object
offer_type            50000 non-null object
price                 50000 non-null object
ab_test               50000 non-null object
vehicle_type          44905 non-null object
registration_year     50000 non-null int64
gear_box              47320 non-null object
power_ps              50000 non-null int64
model                 47242 non-null object
odometer              50000 non-null object
registration_month    50000 non-null int64
fuel_type             45518 non-null object
brand                 50000 non-null object
unrepaired_damage     40171 non-null object
ad_created            50000 non-null object
no_pictures           50000 non-null int64
postal_code           50000 non-null int64
last_seen             50000 non-null object
dtypes: int64(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  no_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Data Exploration and Cleaning

Let's further examine data to find areas where we can clean the data.

In [4]:
autos.describe(include = "all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  no_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

Let's explore the columns that have all or almost all values to be of the same type starting with the columns storing text values.

In [5]:
# Exploring Seller column #
print(autos["seller"].value_counts())
print(autos["seller"].head(10))

privat        49999
gewerblich        1
Name: seller, dtype: int64
0    privat
1    privat
2    privat
3    privat
4    privat
5    privat
6    privat
7    privat
8    privat
9    privat
Name: seller, dtype: object


In [6]:
# Exploring offer_type column #
print(autos["offer_type"].value_counts())
print(autos["offer_type"].head(10))

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64
0    Angebot
1    Angebot
2    Angebot
3    Angebot
4    Angebot
5    Angebot
6    Angebot
7    Angebot
8    Angebot
9    Angebot
Name: offer_type, dtype: object


We can see that seller and offer_type columns have values of almost the same type. These columns can be dropped as they don't have any useful information.

In [7]:
# Drop seller and offer_type columns #
autos = autos.drop(["seller", "offer_type"], axis =1)

Let's explore the remaining columns to see whether they have values of only one type.


In [8]:
# Exploring no_pictures column #
print(autos["no_pictures"].value_counts())
print(autos["no_pictures"].head(10))

0    50000
Name: no_pictures, dtype: int64
0    0
1    0
2    0
3    0
4    0
5    0
6    0
7    0
8    0
9    0
Name: no_pictures, dtype: int64


The no_pictures column contains only zeros and hence it can be dropped as well.

In [9]:
# Drop the no_pictures column #
autos.drop("no_pictures", axis = 1)


date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine

The price and odometer columns have numerical values stored as text. We will clean these two columns and convert them to numeric types.

In [10]:
# cleaning price column #
autos["price"] = (autos["price"].str.replace("$", "").str.replace(",", "").astype(int))
autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
5         7900
6          300
7         1990
8          250
9          590
10         999
11         350
12        5299
13        1350
14        3999
15       18900
16         350
17        5500
18         300
19        4150
20        3500
21       41500
22       25450
23        7999
24       48500
25          90
26         777
27           0
28        5250
29        4999
         ...  
49970    15800
49971      950
49972     3300
49973     6000
49974        0
49975     9700
49976     5900
49977     5500
49978      900
49979    11000
49980      400
49981     2000
49982     1950
49983      600
49984        0
49985     1000
49986    15900
49987    21990
49988     9550
49989      150
49990    17500
49991      500
49992     4800
49993     1650
49994     5000
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [11]:
# cleaning the odometer column #
autos["odometer"] = (autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int))
autos.rename({"odometer": "odometer_km"}, axis = 1, inplace = True)
autos["odometer_km"]


0        150000
1        150000
2         70000
3         70000
4        150000
5        150000
6        150000
7        150000
8        150000
9        150000
10       150000
11       150000
12        50000
13       150000
14       150000
15        80000
16       150000
17       150000
18       150000
19       150000
20       150000
21       150000
22        10000
23       150000
24        30000
25       150000
26       125000
27       150000
28       150000
29       150000
          ...  
49970     60000
49971    150000
49972    150000
49973    150000
49974    150000
49975    100000
49976    150000
49977    150000
49978    150000
49979     70000
49980    125000
49981    150000
49982     90000
49983    150000
49984    150000
49985    150000
49986    125000
49987     50000
49988    150000
49989    150000
49990     30000
49991    150000
49992    125000
49993    150000
49994    150000
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Lengt

## Exploring Price and Odometer columns

In [12]:
# odometer column #
print(autos["odometer_km"].unique().shape)
print(autos["odometer_km"].describe())


(13,)
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


In [13]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The odometer column has a maximum value of 150000 and a minimum value of 5000. All the values are rounded and there are more high mileage vehicles when compared to low mileage ones.

In [14]:
# price column #
print(autos["price"].unique().shape)
print(autos["price"].describe())

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [15]:
autos["price"].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
8              1
9              1
10             7
11             2
12             3
13             2
14             1
15             2
17             3
18             1
20             4
25             5
29             1
30             7
35             1
40             6
45             4
47             1
49             4
50            49
55             2
59             1
60             9
65             5
66             1
            ... 
151990         1
155000         1
163500         1
163991         1
169000         1
169999         1
175000         1
180000         1
190000         1
194000         1
197000         1
198000         1
220000         1
250000         1
259000         1
265000         1
295000         1
299000         1
345000         1
350000         1
999990         1
999999         2
1234566        1
1300000        1
3890000        1
10000000       1
11111111       2
12345678      

There are totally 2357 values in the price column and all the values are rounded. There are 1421 values having price of \$0
and a maximum value of 100 million dollars which is highly unlikely.

We keep the 1 dollar values as they could be the starting prices at the auction and we will remove the values above 350000 dollars as the price values raise steadily untill this value and jumps to less realistic values.


In [16]:
autos = autos[autos["price"].between(1,351000)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring Date Columns

The following are the five columns that represent date values:
- date_crawled
- ad_created
- registration_month
- registration_year
- last_seen

The columns date_crawled, last_seen were added by the crawler and registration_month, registration_year, ad_created are from the actual website.


Let's start exploring these columns in more detail

In [17]:
print(autos["date_crawled"].head(5))
print("\n")
print(autos["last_seen"].head(5))
print("\n")
print(autos["ad_created"].head(5))
print("\n")
print(autos["registration_month"].head(5))
print("\n")
print(autos["registration_year"].head(5))

0    2016-03-26 17:47:46
1    2016-04-04 13:38:56
2    2016-03-26 18:57:24
3    2016-03-12 16:58:10
4    2016-04-01 14:38:50
Name: date_crawled, dtype: object


0    2016-04-06 06:45:54
1    2016-04-06 14:45:08
2    2016-04-06 20:15:37
3    2016-03-15 03:16:28
4    2016-04-01 14:38:50
Name: last_seen, dtype: object


0    2016-03-26 00:00:00
1    2016-04-04 00:00:00
2    2016-03-26 00:00:00
3    2016-03-12 00:00:00
4    2016-04-01 00:00:00
Name: ad_created, dtype: object


0    3
1    6
2    7
3    6
4    7
Name: registration_month, dtype: int64


0    2004
1    1997
2    2009
3    2007
4    2003
Name: registration_year, dtype: int64


We can observe that the values in the columns date_crawled, last_seen, ad_created is stored as strings and the values in the columns registration_month, registration_year is stored as integers.

Let's calculate the distribution of the values in these columns

In [18]:
autos[["date_crawled", "last_seen", "ad_created"]][:10]

date_crawled            last_seen           ad_created
0  2016-03-26 17:47:46  2016-04-06 06:45:54  2016-03-26 00:00:00
1  2016-04-04 13:38:56  2016-04-06 14:45:08  2016-04-04 00:00:00
2  2016-03-26 18:57:24  2016-04-06 20:15:37  2016-03-26 00:00:00
3  2016-03-12 16:58:10  2016-03-15 03:16:28  2016-03-12 00:00:00
4  2016-04-01 14:38:50  2016-04-01 14:38:50  2016-04-01 00:00:00
5  2016-03-21 13:47:45  2016-04-06 09:45:21  2016-03-21 00:00:00
6  2016-03-20 17:55:21  2016-03-23 02:48:59  2016-03-20 00:00:00
7  2016-03-16 18:55:19  2016-04-07 03:17:32  2016-03-16 00:00:00
8  2016-03-22 16:51:34  2016-03-26 18:18:10  2016-03-22 00:00:00
9  2016-03-16 13:47:02  2016-04-06 10:46:35  2016-03-16 00:00:00

In [19]:
# Exploring date crawled column #
(autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_index())

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [20]:
(autos["date_crawled"].str[:10].value_counts(normalize = True, dropna = False).sort_values())

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

It looks like the website was crawled for roughly a month from march to april 2016. The distribution of the data crawled each day is roughly uniform.

In [21]:
# Exploring Last seen column #
(autos["last_seen"].str[:10].value_counts(normalize = True, dropna = False).sort_index())

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The last_seen column throws some light on when the crawler last saw the ad which allows us to determine when the ad was removed, presumably because the car was sold.

Th distribution is roughly uniform except the last few days where the values are 4-5 times more than the previous days. This might be related to the data crawled and they don't indicate car sales.

In [22]:
(autos["ad_created"].str[:10].value_counts(normalize = True, dropna = False).sort_index())

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

There are variety of ad creation dates. Most of the ads fall into 1 month category. The values range from June 2015 to April 2016

In [23]:
# Exploring registration_year column #
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The registration_year column values indicate the age of the car. But, there are some odd values in the column. The maximum value is 9999 which goes into future and the minimum value is 1000 which goes back to past, long before cars were invented.

## Dealing with Incorrect Registration Year Data

Because a car cannot be registered after the ad was seen, any vehicle with a registration year above 2016 is inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

Let's see what percentage of values fall into this range.

In [24]:
(~autos["registration_year"].between(1900,2016)).sum()/autos.shape[0]

0.038793369710697

Roughly 4% of the data fall into this range. So it's ok to remove this data.

In [25]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [26]:
(autos["registration_year"].value_counts(normalize = True).sort_values(ascending = False).head(10))

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

We can see that most of the vehicles listed on the website were registered in the past 15-20 years.

## Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. We will use aggregation to understand the brand column.

In [27]:
autos["brand"].value_counts().sort_values(ascending = False)

volkswagen        9862
bmw               5137
opel              5022
mercedes_benz     4503
audi              4041
ford              3263
renault           2201
peugeot           1393
fiat              1197
seat               853
skoda              766
nissan             713
mazda              709
smart              661
citroen            654
toyota             593
hyundai            468
sonstige_autos     458
volvo              427
mini               409
mitsubishi         384
honda              366
kia                330
alfa_romeo         310
porsche            286
suzuki             277
chevrolet          266
chrysler           164
dacia              123
daihatsu           117
jeep               106
subaru             100
land_rover          98
saab                77
jaguar              73
daewoo              70
trabant             65
rover               62
lancia              50
lada                27
Name: brand, dtype: int64

German manufacturers represent four out of the top five brands and the most popular brand is Volkswagen followed by BMW and Opel. There are lots of car brands having a significant number of listings but, we will limit our analysis to brands having 5% of the total listings.

In [28]:
no_brands = autos["brand"].value_counts(normalize = True).sort_values(ascending = False)
#print(no_brands)
top_brands = no_brands[no_brands > 0.05].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [29]:
brand_mean_price = dict()
for brand in top_brands:
    brand_val = autos[autos["brand"] == brand]
    mean_val = brand_val["price"].mean()
    brand_mean_price[brand] = int(mean_val)

brand_mean_price



{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

There is a distinct gap in price between the brands. Audi, BMW and Mercedes Benz are expensive. Ford and opel are less expensive.Volkswagen is in the middle range - This might be the reason for it's popularity.

## Calculating Mileage For Top Car Brands

We will calculate average mileage for the top car brands to see if there is any connection with mean price.

In [30]:
# Converting brand_mean_price dictionary to a dataframe #
mean_price = pd.Series(brand_mean_price).sort_values(ascending = False)
mean_price_df = pd.DataFrame(mean_price, columns=["mean_price"])
mean_price_df

mean_price
audi                 9336
mercedes_benz        8628
bmw                  8332
volkswagen           5402
ford                 3749
opel                 2975

In [31]:
# calculating average mileage for top brands #
brand_mean_mileage = dict()
for brand in top_brands:
    brand_val = autos[autos["brand"] == brand]
    mean_val = brand_val["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_val)

brand_mean_mileage

{'audi': 129157,
 'bmw': 132572,
 'ford': 124266,
 'mercedes_benz': 130788,
 'opel': 129310,
 'volkswagen': 128707}

In [32]:
# Converting brand_mean_mileage dictionary to a dataframe #
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending = False)
mean_mileage_df = pd.DataFrame(mean_mileage, columns=["mean_mileage"])
mean_mileage_df



mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [33]:
# Creating a new dataframe with mean_price and mean_mileage #
brand_mean_price = pd.DataFrame(mean_mileage, columns = ["mean_mileage"])
brand_mean_price["mean_price"] = mean_price
brand_mean_price

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

Based on the above data, there are a few observations:
- Car mileages do not vary as much as price do with car brands.
- Expensive cars have more mileage when compared to less expensive ones.


## Conclusion

In this project, we cleaned and analyzed used car listings data and found that German manufacturers represent four out of the top five car brands, almost fifty percent of the total listings. We also explored the relationship between the price and mileage of top car brands and concluded that expensive cars have more mileage compared to less expensive ones.